tests/fixtures-valid.js

```diff
@@ -5,6 +5,32 @@ const path = require(`path`);
 const chalk = require(`chalk`);
 const Ajv = require(`ajv`);
 
+// interactive commandline support
+const minimist = require(`minimist`);
+
+const args = minimist(process.argv.slice(2), {
+  boolean: [`h`],
+  string: [`f`], // convert to file with possibly completion?
```

Comment:
Actually, I think it would be better to make use of the auto-completion for files in terminal. The most common use case will be to just check a newly added fixture and this can be quite quick with file completion.

Look at the [cli/export-fixture.js](https://github.com/OpenLightingProject/open-fixture-library/blob/master/cli/export-fixture.js) script, where fixtures are added using paths without a parameter at the end of the command:

```
$ node cli/export-fixture.js -h
Usage: /home/felix/Development/open-fixture-library/cli/export-fixture.js -p <plugin name> [ -a | <fixture> [<more fixtures>] ]
Options:
  --plugin,       -p: Which plugin should be used to export fixtures.
                      E. g. ecue or qlcplus
  --all-fixtures, -a: Use all fixtures from register
  --output-dir,   -o: If set, save outputted files in this directory
                      instead of printing the contents in the console
  --help,         -h: Show this help message.
$ node cli/export-fixture.js -p ecue fixtures/generic/desk-channel.json fixtures/generic/rgb-fader.json 
```

We could use a similar approach for fixture-valid as well. It may be worth a discussion whether we should allow `-a` as an option (this results in an error if neither `-a` nor specific fixtures are given) or use it as default. We should use the same method in cli/export-fixture then (that's a task for #439). Furthermore, it may be considerable to also allow a directory instead of a fixture file; this will check all JSON files in that directory (not recursively). For instance:

```bash
$ node tests/fixtures-valid.js # all fixtures
$ node tests/fixtures-valid.js fixtures/generic/desk-channel.json fixtures/generic/rgb-fader.json # only two generic fixtures
$ node tests/fixtures-valid.js fixtures/generic/ # all generic fixtures
```

Reply:
That's what I tried to do initially as well because that's the most idiomatic way, but that would require to refactor the entire fixture_valid.js which I am simply able to do currently because of time constraints.

Reply:
Hm, I think the problem is that you can't (always) guess manKey and fixKey from a file path, right? And these keys are used in fixture-valid for uniqueness checks.

Right, one could disable uniqueness checks in this case in fixture-valid, but I totally understand that you don't want to touch this..

So let's just keep the manKey/fixKey syntax you proposed. I still have to test the code on my computer, but I think we can merge it then.

Reply:
> Hm, I think the problem is that you can't (always) guess manKey and fixKey from a file path, right? And these keys are used in fixture-valid for uniqueness checks.

Exactly, the current code makes too many assumptions in the wrong places. Overall, the entire code is mostly procedural (probably because of the anemic domain model) instead of making use the OOP practices javascript enables us (at least IMHO).

Reply:
Thanks @Swiftb0y for this improvement!

A problem with the way you're specifying fixtures in the commandline is that it uses the semicolon character `;`, since that is interpreted by bash (and probably other shells) as a separator between program calls. That means that calling `./tests/fixtures-valid.js -f american-dj;adb` (test all American DJ and all ADB fixtures) actually only tests all ADJ fixtures and then tries to call `adb` (which happens to be the [Android Debug Bridge](https://developer.android.com/studio/command-line/adb) program). You'd have to explicitly pass a string: `./tests/fixtures-valid.js -f "american-dj;adb"`.

Thus, I'd also suggest passing the fixtures as `argv._` (i.e. like this: `./test/fixtures-valid.js fixtures/american-dj/* fixtures/adj/*`). That way, the shell handles autocomplete and glob syntax without any additional work. You don't need to change `fixture_valid.js` for that, since you can deduct `manKey` and `fixKey` from the path.

We could still allow the `manKey/fixKey` shortcut (just check then if the string contains `.json`).

Could you give it a try? Otherwise just tell me, then I'll do it.

Reply:
I used a semicolon as a separator because the initial idea was to pass everything via a single flag. Using the excess arguments seems like a good Idea but in that case I'd change the program to do nothing when passed nothing and just add a `-a`/`--all-fixtures` flag. However, I imagine that this is a breaking change and would have to be documented somewhere (please tell me where). 

Even though the autocomplete functionality would be useful, in my opinion, its not a good idea because it suggests that you can pass arbitrary files/paths while that is not the case. If we really want to enable autocomplete for frequent users, I'd much rather use [yargs](http://yargs.js.org/) and then use its [autocompletion functionality](http://yargs.js.org/docs/#api-completioncmd-description-fn).

Reply:
Doing nothing without specified fixtures and relying on `-a` / `--all-fixtures` seems like a good idea. That's the same way we specify fixtures for the `run-export-test.js`, `export-fixture.js` and `diff-plugin-outputs.js` CLI scripts. I don't think it implies that arbitrary files are supported, and if someone tries, a helpful error message can be thrown.

Since the tests and CLI scripts are not public facing API, we don't really need to care about breaking changes. For people accustomed to the current behavior, we can show the help text without specified arguments.

The completion feature of yargs looks nice but I think it's not suitable in our case. We want to autocomplete manufacturer and fixture keys, which are regularly added. So users would have to keep updating their shell configuration to include newly added fixtures, while default shell file completion is for free.

Reply:
> Since the tests and CLI scripts are not public facing API, we don't really need to care about breaking changes. For people accustomed to the current behavior, we can show the help text without specified arguments.

Ok I was just mentioning that because that's always something to care about when working on bigger projects like Mixxx. 

> The completion feature of yargs looks nice but I think it's not suitable in our case. We want to autocomplete manufacturer and fixture keys, which are regularly added.

This doesn't pose a problem. You need to add the generated completion script to your bashrc/zshrc once and then the completion works by calling the completion function each time. So the completion options are generated on demand. The primary issue is that someone has to do that previous setup before the completion works.

If you still favor your approach, I can implement it even though I do not think its the most idiomatic.

Reply:
Thinking about how to handle breaking changes is never a bad idea :)

Ah, then I misunderstood how the yargs completion feature works. I actually do still favor the files approach, since it doesn't require any prior configuration, making it easier for both new contributors and us when we checkout fork repos (like yours :wink:) to test PRs. @fxedel What's your opinion on this?

Reply:
I also prefer the files approach. One could deduct manKey/fixKey like so:

- Convert all given relative paths to absolute paths
- Check if these paths are JSON files in a direct subdirectory of fixtures/ → if not, give an error message
- manKey = directory name; fixKey = filename without ".json"

Actually, that's very much like the [implementation in `cli/export-fixture.js`](https://github.com/OpenLightingProject/open-fixture-library/blob/master/cli/export-fixture.js#L56-L62), just without the check and the user-visible error message.

```js
  fixtures = args._.map(relativePath => {
    const absolutePath = path.join(process.cwd(), relativePath);
    return [
      path.basename(path.dirname(absolutePath)), // man key
      path.basename(absolutePath, path.extname(absolutePath)) // fix key
    ];
  });
```

In fact, it *could* also be allowed to test a fixture at a very different location, as long as its directory name is a valid manufacturer key. If we allowed this, we'd have to check whether the directory name is valid (instead of whether the file is in a subdirectory of fixtures/).

Reply:
Oh, and auto completion with yargs sounds pretty cool. Especially the `--plugin` parameter in `cli/export-fixture.js` would benefit from that!

Reply:
> Oh, and auto completion with yargs sounds pretty cool. Especially the --plugin parameter in cli/export-fixture.js would benefit from that!

the issue with the yargs autocomplete is that (afaik) it has to be setup manually for every single single file. So in case we want to create a better UX for the CLI interaction, IMO it would rather make more sense to refactor a bunch of the code into less coupled libraries and then create a single CLI application that works with subcommands (similar to git). 

In conclusion:
I am fine with implementing that 'pseudo-file' implementation because its just supposed to be a temporary workaround anyways (until `fixture-valid.js` was refactored to allow passing checking arbitrary files without the need of the `fixKey` or `manKey`)